### PR TITLE
#93 Extract interfaces for single and composite nodes

### PR DESCRIPTION
**Pull request by tomaszmichalak**
Extract `single` and `composite` node interfaces.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] My code follows the [code style](/CONTRIBUTING.md#coding-conventions) of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

---
I hereby agree to the terms of the Knot.x Contributor License Agreement.

### CODE REVIEW

File: handler/core/src/main/java/io/knotx/fragments/task/factory/node/subtasks/SubtasksNodeFactory.java [hunk 1]
```diff
@@ -52,8 +53,27 @@ public Node initNode(GraphNodeOptions nodeOptions, Map<String, Node> edges,
...
+        return filter(transition).map(edges::get);
+      }
+
+      @Override
+      public List<Node> getNodes() {
+        return nodes;
+      }
+
+      private Optional<String> filter(String transition) {
+        return ERROR_TRANSITION.equals(transition) || SUCCESS_TRANSITION.equals(transition)
```
Comment (Mateusz512) [inline handler/core/src/main/java/io/knotx/fragments/task/factory/node/subtasks/SubtasksNodeFactory.java]: `StringUtils.equalsAny`

Comment (Mateusz512) [reply handler/core/src/main/java/io/knotx/fragments/task/factory/node/subtasks/SubtasksNodeFactory.java]: or
```return Optional.of(transition)
.filter(tr -> StringUtils.equalsAny(tr, ERROR_TRANSITION, SUCCESS_TRANSITION);
```

File: handler/engine/src/test/java/io/knotx/fragments/engine/Nodes.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    return node;
+  }
+
+  static CompositeNode composite(String nodeId, List<Node> rootNodes, Node success, Node error) {
+    CompositeNode node = composite(nodeId, rootNodes, success);
+    when(node.next(matches(ERROR_TRANSITION))).thenReturn(Optional.ofNullable(error));
+    return node;
+  }
+
+  static SingleNode node(String nodeId, Function<FragmentContext, Single<FragmentResult>> function) {
```
Comment (Skejven) [inline handler/engine/src/test/java/io/knotx/fragments/engine/Nodes.java]: To be consistent: `node` -> `single` ?

Comment (tomaszmichalak) [reply handler/engine/src/test/java/io/knotx/fragments/engine/Nodes.java]: Sure, we can change this value. However, the `SingleNode` name is the best one - when you check the interface you will see that this node allows modifying a Fragment. It is more like `FunctionNode` or `ProcessingNode`. `CompositeNode` is a special node that exposes its internal complexity to Engine. But still, we can implement `SingleNode` that would work like `CompositeNode`.

Comment (Skejven) [reply handler/engine/src/test/java/io/knotx/fragments/engine/Nodes.java]: Ok, I was not precise. I meant method name.

Comment (tomaszmichalak) [reply handler/engine/src/test/java/io/knotx/fragments/engine/Nodes.java]: Fixed, however still `single` is not good name:
```
interface SingleNode extends Node {

  Single<FragmentResult> execute(FragmentContext fragmentContext);

}
```

**Pull request merged**